include/odp/api/spec/packet.h

```diff
@@ -1378,12 +1394,44 @@ uint32_t odp_packet_l4_offset(odp_packet_t pkt);
 int odp_packet_l4_offset_set(odp_packet_t pkt, uint32_t offset);
 
 /**
+ * Layer 3 checksum check status
+ *
+ * Returns the result of the latest layer 3 checksum check done for the packet.
+ * The status tells if checksum check was attempted and the result of the
+ * attempt. It depends on packet input (or IPSEC) configuration, packet content
+ * and implementation capabilities if checksum check is attempted for a packet.
+ *
+ * @param pkt     Packet handle
+ *
+ * @return L3 checksum check status
+ */
+odp_packet_chksum_status_t odp_packet_l3_chksum_status(odp_packet_t pkt);
```

Comment:
Instead of this I would prefer to have in ./include/odp/api/spec/packet_flags.h  / ./platform/linux-generic/odp_packet_flags.c something like:

int odp_packet_has_l3_csum(odp_packet_t pkt);
int odp_packet_has_l4_csum(odp_packet_t pkt);

* @retval non-zero operation was performed
* @retval 0 operation was not performed

For completeness, we may add explicit csum operation result call but likely odp_packet_has_error() or odp_packet_has_l3_error() or odp_packet_has_l4_error() will be preferred.

int odp_packet_has_l3_csum_error(odp_packet_t pkt);
int odp_packet_has_l4_csum_error(odp_packet_t pkt);

Reply:
Why you'd prefer two calls instead of one? With ABI compat, the number of function calls per packet  matters. Also, "has checksum been checked" and "checksum check result" are likely reading the same bits in packet header anyway. So, it makes sense to do both checks in one go.

Reply:
There are 3 cases:

1.	Csum was done + success 
2.	Csum was done + error
3.	Csum was not done


First thing you will usually do with a packet (for N number of reasons):

If (odp_packet_has_error())
	odp_packet_free(pkt);

This will remove packets with “Csum was done + error”.
The only question remaining: “Csum was done + success” or “Csum was not done” 

More, the implementation will likely put this info in odp_packet_hdr(pkt)->p.input_flags, so it makes sense to look like the other APIs from there.

Reply:
There are actually more than 3 cases when we think about the usage of the result. Basically for each 3 cases you list, there are 2 subcases: a) The application wants to know that the checksum is correct (i.e. it actually receives the packet) b) The application does not care about the correctness of the checksum either because the packet gets dropped anyway for other reasons or because it is L4 checksum and the packet is going to be forwarded and should not be dropped even if the L4 checksum is not correct.

So, with checksum offload enabled, an application should not use odp_packet_has_error() for forwarded packets but odp_packet_has_l2_error() and odp_packet_has_l3_error() instead.

Reply:
odp-linux implementation will add two bits into p.input_flags (checked: 0/1, correct: 0/1), other implementations read/check something else from their packet descriptor.

Reply:
Actually, “correct: 0/1” already exists:
error_flags.ip_err 
error_flags.tcp_err 
error_flags.udp_err 

And existing API: 
int odp_packet_has_l3_error(odp_packet_t pkt);
int odp_packet_has_l4_error(odp_packet_t pkt);

What is needed (from my point in view) is “checked: 0/1”:

int odp_packet_has_l3_csum(odp_packet_t pkt);
int odp_packet_has_l4_csum(odp_packet_t pkt);

I suggest to go with this for a cleaner API.

Anyway, one function call or two is less relevant for a ‘real’ application: only if csum was done or not is really important. The ‘cost’ for the extra call is not significant and can be avoided.

Reply:
With my proposal application may always reduce the number of return values by itself:
static inline has_l3_csum(odp_packet_t pkt) {
    return odp_packet_l3_chksum_status(pkt) != ODP_PACKET_CHKSUM_UNKNOWN;
}

Also, sometimes application may want to know which error(s) packet has. E.g. it may not tolerate length errors, but  a checksum error may be OK for some packets.

A non-inlined function call is about 7-8 CPU cycles. L2fwd cycle count is around 130-250 cycles (no-copy vs copy) per packet (non ABI compat). A real application could be around 500 cycles per packet (ABI compat). That's about -1.5% max packet rate for every extra function call.

Reply:
Fine.
Implementation for dpdk pktio?